planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+    .AggregateByEventData<TEvent>(Expression<Func<TEvent, int>> expression) // what's the identity of the projected, aggregate doc?
+
+    .DeleteOn<TEvent>() // the aggregate document should be deleted if an event of this type hits
+                        // In this case, we could look ahead and "know" in the rebuilds not to rebuild
+                        // a projection for this aggregate as an optimization
+
+    .SnapshotInIncrementsOf(10) // Then maybe there are other options to retain old snapshots or not? Dunno
+    
+```
+
+
+Now that folks are used to the convention based signature approach of `StartUp` in ASP.Net Core, I suggest doing something
+similar for the projections like so:
+
+```
+    // This purposely does not implement any kind of interface
+    // Defines how to apply changes to one kind of projected document
+    public class SampleTransformer
+    {
+        
+
+        // Support constructor injection here for services from the container?
+        // Could also support method injection too, w/ maybe something like the ASP.net Core
+        // [FromServices] attribute on parameters???
+
+        // Marten should conform to the arguments of the method, so any combination of
+        // the actual event (either the specific type or a base type), the event wrapper
+        // with metadata, and the stream model. Plus the projected document where appropriate
+
+        // These may not be necessary, but if so, this is to 
+        // get at what the identity of the aggregate document is
+        // Prefer the FI version of this though
+        public int AggregateBy(Event1 @event);
+        public int AggregateBy(EventWrapper @event);
+
+        // Sync version with raw event
+        public void Apply(Event1 @event, ProjectedDocument doc);
+
+        // Sync version with event wrapper, could be a custom implementation of IEvent
+        // to get at event metadata & stream metadata or partitioning information
+        public void Apply(EventWrapper<Event1> @event, MyEventStream stream, ProjectedDocument doc);
+
+        // Sync version w/ immutable projected docs
+        public ProjectedDocument Apply(Event1 @event, ProjectedDocument doc);
+
+        // apply a partial update to an aggregate document, but this wouldn't
+        // get used if there's some reason to fetch the whole aggregate for a segment
+        // of events. 
+        public void Partial(Event1, @event, IDocumentSession session);
+
+        // Should the aggregate document be deleted based on some event data?
+        public bool ShouldBeDeleted(Event1 @event);
+        public bool ShouldBeDeleted(EventWrapper @event);
+        public bool ShouldBeDeleted(EventWrapper @event, MyEventStream stream);
+    }
+```
+
+#### Defining a Projection to a SQL Statement(s)
+
+TODO
+
+#### Inline Projections
+
+There's been some complaints and obvious confusion about the `IEvent` interface and how the metadata properties like
```

Comment:
I think that it's not only in the inline but in general (as far as I understood).

Reply:
Correct. Only a problem for the inline projections, but we might as well close that down. I think it dovetails nicely with the tombstone idea